tests/meltano/core/plugin/singer/test_tap.py

```diff
@@ -201,44 +201,45 @@ def mock_discovery(*args, **kwargs):  # noqa: ARG001
-                # Apply catalog rules to store the cache key
-                subject.apply_catalog_rules(invoker)
-                assert catalog_cache_key_path.exists()
-
-                # If the cache key hasn't changed, discovery isn't invoked again
-                mocked_run_discovery.reset_mock()
-                await subject.discover_catalog(invoker)
-
-                mocked_run_discovery.assert_not_called()
-                assert json.loads(catalog_path.read_text()) == {"discovered": True}
-                assert catalog_cache_key_path.exists()
-
-                # If the cache key no longer matches, discovery is invoked again
-                mocked_run_discovery.reset_mock()
-                catalog_cache_key_path.write_text("bogus")
-                await subject.discover_catalog(invoker)
-
-                assert mocked_run_discovery.called
-                assert json.loads(catalog_path.read_text()) == {"discovered": True}
-                assert not catalog_cache_key_path.exists()
+            with mock.patch.object(SingerTap, "warn_property_not_found"):
+                with mock.patch.object(
+                    SingerTap,
+                    "run_discovery",
+                    side_effect=mock_discovery,
+                ) as mocked_run_discovery:
+                    await subject.discover_catalog(invoker)
+
+                    assert mocked_run_discovery.called
+                    assert json.loads(catalog_path.read_text()) == {"discovered": True}
+                    assert not catalog_cache_key_path.exists()
+
+                    # If there is no cache key, discovery is invoked again
+                    mocked_run_discovery.reset_mock()
+                    await subject.discover_catalog(invoker)
+
+                    assert json.loads(catalog_path.read_text()) == {"discovered": True}
+                    assert not catalog_cache_key_path.exists()
+
+                    # Apply catalog rules to store the cache key
+                    subject.apply_catalog_rules(invoker)
+                    assert catalog_cache_key_path.exists()
+
+                    # If the cache key hasn't changed, discovery isn't invoked again
+                    mocked_run_discovery.reset_mock()
+                    await subject.discover_catalog(invoker)
+
+                    mocked_run_discovery.assert_not_called()
+                    assert json.loads(catalog_path.read_text()) == {"discovered": True}
+                    assert catalog_cache_key_path.exists()
+
+                    # If the cache key no longer matches, discovery is invoked again
+                    mocked_run_discovery.reset_mock()
+                    catalog_cache_key_path.write_text("bogus")
+                    await subject.discover_catalog(invoker)
+
+                    assert mocked_run_discovery.called
+                    assert json.loads(catalog_path.read_text()) == {"discovered": True}
+                    assert not catalog_cache_key_path.exists()
 
```

Comment:
You could add something like this to test the output:

```suggestion

    @pytest.mark.parametrize(
        ("rule", "catalog", "message", "emit_warning"),
        (
            pytest.param(
                MetadataRule("foo", ["properties", "bar"], "baz", False),
                {"streams": []},
                "Stream `foo` was not found",
                True,
                id="stream_not_found",
            ),
            pytest.param(
                MetadataRule("foo", ["properties", "bar"], "baz", False),
                {"streams": [{"tap_stream_id": "foo"}]},
                "Property `foo.bar` was not found",
                True,
                id="property_not_found",
            ),
            pytest.param(
                MetadataRule("foo", ["properties", "bar"], "baz", False),
                {"streams": [{"tap_stream_id": "foo"}]},
                "Stream `foo` was not found",
                False,
                id="stream_is_found",
            ),
            pytest.param(
                MetadataRule("foo", ["properties", "bar"], "baz", False),
                {
                    "streams": [
                        {
                            "tap_stream_id": "foo",
                            "schema": {"properties": {"bar": {"type": "string"}}},
                        },
                    ],
                },
                "Property `foo.bar` was not found",
                False,
                id="property_is_found",
            ),
            pytest.param(
                MetadataRule("*", ["properties", "bar"], "baz", False),
                {
                    "streams": [],
                },
                "Stream `foo` was not found",
                False,
                id="stream_is_wildcard",
            ),
            pytest.param(
                MetadataRule("foo", ["properties", "*"], "baz", False),
                {"streams": [{"tap_stream_id": "foo"}]},
                "Property `foo.bar` was not found",
                False,
                id="property_is_wildcard",
            ),
        ),
    )
    def test_warn_property_not_found(
        self,
        subject: SingerTap,
        capsys: pytest.CaptureFixture,
        rule: MetadataRule,
        catalog: dict,
        message: str,
        emit_warning: bool,
    ):
        subject.warn_property_not_found([rule], catalog)

        # If the warning is emitted, it should be in the output
        assert emit_warning is (message in capsys.readouterr().out)

```

Reply:
Thanks for this suggestion. I'll play with it after work tomorrow. I appreciate all the suggestions & applied the others!

Reply:
when I run `pytest`, the following errors are returned.

```
SKIPPED [1] tests/meltano/cli/test_elt.py:192: Test is only for Windows
XFAIL tests/meltano/cli/test_add.py::TestCliAdd::test_add_fails - Uninstall not implemented yet.
FAILED tests/meltano/core/plugin/singer/test_tap.py::TestSingerTap::test_warn_property_not_found[property_not_found] - AssertionError: assert True is ('Property `bar` was not found in the schema of stream `foo`' in '')
FAILED tests/meltano/core/plugin/singer/test_tap.py::TestSingerTap::test_warn_property_not_found[stream_not_found] - AssertionError: assert True is ('Stream `foo` was not found' in '')
```

however, when the individual test is run, `poetry run pytest tests/meltano/core/plugin/singer/test_tap.py::TestSingerTap::test_warn_property_not_found`, all 6 versions pass.

I'm not sure what's happening here, and I'm not that familiar with `pytest`. Do you have any suggestions for this error?

Reply:
The issue above has been resolved, it seems doing `poetry run pytest` ends up with the log captured in the built in pytest fixture `caplog`, whereas if the test is run by itself, the output is not captured in `caplog`, but instead in `capsys`. i'm joining the output from both and testing for the presence of the expected warning messages.